repose-aggregator/components/services/atom-feed-service/atom-feed-service-impl/src/main/scala/org/openrepose/nodeservice/atomfeed/impl/AtomEntryStreamBuilder.scala

```diff
@@ -43,67 +46,84 @@ object AtomEntryStreamBuilder {
-        Option(Await.result(connectionFuture, authenticationTimeLimit))
+        val connectionFuture = Future(arf.authenticateRequest(feedReadRequest, context))
+        Option(Await.result(connectionFuture, authenticationTimeout))
       case None =>
-        Some(baseFeedConnection)
+        Some(feedReadRequest)
     }
 
-    authenticatedConnection match {
-      case Some(urlConnection) =>
+    authenticatedRequest match {
+      case Some(readRequest) =>
         val tracingHeader = TracingHeaderHelper.createTracingHeader(context.getRequestId, "1.1 Repose (Repose/" + context.getReposeVersion + ")", None)
-        urlConnection.setRequestProperty(CommonHttpHeader.TRACE_GUID.toString, tracingHeader)
+        feedReadRequest.getHeaders.put(CommonHttpHeader.TRACE_GUID.toString, util.Arrays.asList(tracingHeader))
 
+        val httpGet = new HttpGet(feedReadRequest.getURI)
+        feedReadRequest.getHeaders foreach { case (key, values) =>
+          values.foreach(value => httpGet.addHeader(key, value))
+        }
+
+        val httpResponse = httpClient.execute(httpGet)
         try {
-          val feedInputStream = urlConnection.getInputStream
-          val feed = parser.parse[Feed](feedInputStream).getRoot
-
-          feed.getLinks.find(link => link.getRel.equals("next")) match {
-            case Some(nextPageLink) =>
-              feed.getEntries.toStream #::: buildR(nextPageLink.getResolvedHref.toURL, context, authenticator, authenticationTimeLimit)
-            case None =>
-              feed.getEntries.toStream
-          }
-        } catch {
-          case ioe: IOException =>
-            authenticator.foreach(_.onInvalidCredentials())
+          val statusCode = httpResponse.getStatusLine.getStatusCode
+
+          if (statusCode >= 200 && statusCode < 300) {
+            val content = Option(httpResponse.getEntity)
+              .map(_.getContent)
+              .getOrElse(new ByteArrayInputStream(Array.empty[Byte]))
+            val feed = parser.parse[Feed](content).getRoot
+
+            feed.getLinks.find(link => link.getRel.equals("next")) match {
+              case Some(nextPageLink) =>
+                feed.getEntries.toStream #::: buildR(nextPageLink.getResolvedHref.toURI, httpClient, context, authenticator, authenticationTimeout)
+              case None =>
+                feed.getEntries.toStream
+            }
+          } else if (statusCode >= 400 && statusCode < 500) {
+            authenticator.foreach(_.onInvalidCredentials)
             if (firstAttempt) {
               // Inform the authenticator that the request failed and retry once
-              buildR(baseFeedUrl, context, authenticator, authenticationTimeLimit, firstAttempt = false)
+              buildR(feedURI, httpClient, context, authenticator, authenticationTimeout, firstAttempt = false)
             } else {
-              throw UnrecoverableIOException(ioe)
+              throw ClientErrorException
```

Comment:
I don't like losing the original exception, and neither will sonar if it ever figures out how to handle sonar.